### ISSUE TITLE
Enabling inference service logging disables streaming responses

### ISSUE THREAD

**Issue by jerryharrow** [kserve/kserve#3809]
/kind bug

**What steps did you take and what happened:**

If you enable logging, the agent sidecar added breaks streaming responses.   Instead of a streamed responses, the caller only gets any response data after the full response has been created.

**What did you expect to happen:**

The streamed response should start and the caller should see each streamed token as they are generated.

**What's the InferenceService yaml:**

Here is the fully functional LLM inference service using a public docker image that can be used to reproduce the problem.
Normally, we have a custom logger to handle the logging, but I have omitted it to simply this example.

custom-model.yaml:
```
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: custom-model
spec:
  predictor:
   # Adding the logger, introduces the agent sidecar which breaks streaming responses
   # Comment out to get expected behavior.
   logger:
      mode: all
   containers:
    - name: kserve-container
      image: jjharrow/facebook-pt-125m-service
      env:
        - name: BENTOML_HOME
          value: /home/bentoml/bento
      args:
        - openllm
        - start
        - --port
        - "8080"
        - --backend=pt
        - --dtype=float32
        - facebook/opt-125m
      ports:
        - containerPort: 8080
```

kubectl apply -f  custom-model.yaml

Once successfully deployed/started get the service address

`HOST={inferenceServiceAddress}`

You can show the behavior using the script `read-stream.sh` to invoke a streaming inference and show timestamps on the responses.   When streaming properly the first response takes about 3-4 secs for me, with the rest completing over the next 5 seconds or so (depending on how long the response is).   When logging is enabled, instead the first response takes 8 seconds with all the output essentially happening at the same time.

read-stream.sh
```
#/bin/bash
# Usage: read-stream.sh http://custom-model.default.cluster.local
HOST=$1

printf '[%s]\n' "$(date '+%Y-%m-%d %H:%M:%S')"
curl -s $HOST/v1/generate_stream -H 'accept: text/event-stream' -H 'Content-Type: application/json' \
  -d '{"prompt": "What is the meaning of life?"}' \
  | while IFS= read -r line; do printf '[%s] %s\n' "$(date '+%Y-%m-%d %H:%M:%S')" "$line"; done
```

With logging enabled:

```
./read-stream.sh $HOST
[2024-07-19 10:52:49]
[2024-07-19 10:52:57] data: {"prompt":"What is the meaning of life?","finished":false,"outputs":[{"index":0,"text":"\n","token_ids":[50118],"cumulative_logprob":0.0,"logprobs":null,"finish_reason":null}],"prompt_token_ids":[2,2264,16,5,3099,9,301,116],"prompt_logprobs":null,"request_id":"openllm-200254079ce84ebda5980c34bffd0d24"}
...
[2024-07-19 10:52:58] data: [DONE]
```

With logging disabled:
```
./read-stream.sh $HOST
[2024-07-19 10:55:59]
[2024-07-19 10:56:03] data: {"prompt":"What is the meaning of life?","finished":false,"outputs":[{"index":0,"text":"\n","token_ids":[50118],"cumulative_logprob":0.0,"logprobs":null,"finish_reason":null}],"prompt_token_ids":[2,2264,16,5,3099,9,301,116],"prompt_logprobs":null,"request_id":"openllm-7977401451b14bb59c13e92865a91047"}
...
[2024-07-19 10:56:07] data: [DONE]
```

From a python environment you can demonstrate it with the openllm.   It is more visual since you can see each word print one at a time, instead of getting the whole response in a single print.

```
pip install openllm==0.4.44 openllm-client==0.4.44
openllm query --endpoint $HOST hello
hello! i'm a new player too! i just started playing. i just want to know what i can do to improve my game. i have a lot of stuff to learn but i'm not sure if i can really improve my game in a very specific way. i'm also new to this game and i'm learning alot. i just want to know if i can improve my game or if i should just continue playing and play. thanks!

Hello, i'm a new player too! i just started playing. i just want to know what i can do to improve my game. i have a lot of stuff to learn but
```

**Anything else you would like to add:**

Any potential workaround suggestions would be appreciated.

**Environment:**

Reproducible with:
curl -s "https://raw.githubusercontent.com/kserve/kserve/release-0.12/hack/quick_install.sh" | bash

- Cloud Environment: GKE cluster 
- Kubernetes version: (use `kubectl version`):
GKE cluster
Client Version: v1.29.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.10-gke.1075001
Docker Desktop K8s:
Client Version: v1.29.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.1
- OS (e.g. from `/etc/os-release`):
NAME="Container-Optimized OS"
ID=cos
PRETTY_NAME="Container-Optimized OS from Google"
HOME_URL="https://cloud.google.com/container-optimized-os/docs"
BUG_REPORT_URL="https://cloud.google.com/container-optimized-os/docs/resources/support-policy#contact_us"
GOOGLE_METRICS_PRODUCT_ID=26
KERNEL_COMMIT_ID=1a8b1dfd94df6f28729aca0565343c0a9f6906ee
GOOGLE_CRASH_ID=Lakitu
VERSION=109
VERSION_ID=109
BUILD_ID=17800.218.69
--
Also reproduced on MacOs Apple M2 Max, 14.5 (23F79)

**Comment by sivanantha321** [comment]
@jerryharrow Thanks for the detailed report. generate_stream endpoint was an experimental support. It was removed in 0.13 release. You should try with OpenAI chat completion / completion endpoints which is introduced in 0.13.0. I have tested with completion streaming endpoint, It seems to be working fine.

**Comment by romanwozniak** [comment]
Hi @sivanantha321, I can reproduce this issue with `/openai/v1/completions` and `/openai/v1/chat/completion` endpoints with Kserve v0.13.1. 

When the `stream: true` is passed in the prediction request, in both cases (with logger and without logger) the response is streamed in chunks. However, if the `InferenceService` is configured with a logger, the response starts being streamed after the full response sequence is generated by the model. 

Unfortunately, I didn't have a chance to look into the implementation details to understand better why this happens.

**Comment by sivanantha321** [comment]
@romanwozniak Okay, So the response is streamed but only after the full response is received by the logger. Is that right ?

**Comment by romanwozniak** [comment]
@sivanantha321, yes, that's correct.

**Comment by jerryharrow** [comment]
> @jerryharrow Thanks for the detailed report. generate_stream endpoint was an experimental support. It was removed in 0.13 release. You should try with OpenAI chat completion / completion endpoints which is introduced in 0.13.0. I have tested with completion streaming endpoint, It seems to be working fine.

This was just a simple example I could show with specific detail.   As was later noted by @romanwozniak, it isn't generate_stream specific.   With a logger enabled, any streaming interface response happens only after the full response is generate (including with NVIDIA NIM containers).

**Comment by sivanantha321** [comment]
@romanwozniak @jerryharrow I am able to reproduce this. Fix is coming soon ✌️

**Comment by romanwozniak** [comment]
> @romanwozniak @jerryharrow I am able to reproduce this. Fix is coming soon ✌️

Amazing, thank you @sivanantha321 🙌 

On a slightly related topic of logging - the latest nightly version of `huggingfaseserver` (i.e. based on the `master` branch, after the upgrade of `vllm` to `0.5.3.post1`) has stopped logging request data in the `kserve-container` (in-container logs, not via the agent/logger sidecar).


`kserve v0.13.1` logged requests to vllm like this:
```
Received request %s: prompt: %r, params: %s, prompt_token_ids: %s ...
```

But the nightly version is no longer producing such log statements. I checked `vllm` to find if any change in behaviour has been made from vllm `v0.4.3` to `v0.5.3.post1`, but it doesn't seem that anything on the vllm part has changed (logging on vllm side us done by https://github.com/vllm-project/vllm/blob/v0.5.3.post1/vllm/entrypoints/logger.py and is controlled by `--disable-log-requests` flag).

If this change is unexpected, I will log a separate issue so that it can be looked in details

**Comment by sivanantha321** [comment]
> > @romanwozniak @jerryharrow I am able to reproduce this. Fix is coming soon ✌️
> 
> Amazing, thank you @sivanantha321 🙌
> 
> On a slightly related topic of logging - the latest nightly version of `huggingfaseserver` (i.e. based on the `master` branch, after the upgrade of `vllm` to `0.5.3.post1`) has stopped logging request data in the `kserve-container` (in-container logs, not via the agent/logger sidecar).
> 
> `kserve v0.13.1` logged requests to vllm like this:
> 
> ```
> Received request %s: prompt: %r, params: %s, prompt_token_ids: %s ...
> ```
> 
> But the nightly version is no longer producing such log statements. I checked `vllm` to find if any change in behaviour has been made from vllm `v0.4.3` to `v0.5.3.post1`, but it doesn't seem that anything on the vllm part has changed (logging on vllm side us done by https://github.com/vllm-project/vllm/blob/v0.5.3.post1/vllm/entrypoints/logger.py and is controlled by `--disable-log-requests` flag).
> 
> If this change is unexpected, I will log a separate issue so that it can be looked in details

~Is this log comes from vLLM ?~ Please raise an issue for this.
